RTL support

SamyPesse opened the pull request:
This PR fixes #134 by adding support for RTL editing. it support editor contents with a mixture of LTR and RTL text.

The functionality is based on DraftJS implementation:

```
To that end, Draft uses a bidi algorithm to determine appropriate text alignment and direction on a per-block basis.

Text is rendered with an LTR or RTL direction automatically as the user types. You should not need to do anything to set direction yourself.
```

### TODOs

- [x] Accept `textAlignment` prop in `Editor`
- [x] Get text direction on a per-block basis (using [direction](https://github.com/wooorm/direction))
- [x] Update `directionMap` when content changed
- [x] Add `dir="rtl"` when required.
- [x] Fix tests

SamyPesse commented:
@ianstormtaylor This is working, but here are some notes:

- The implementation is based on the same idea as Draft, where `directionMap` is [updated](https://github.com/facebook/draft-js/blob/1ce48f67728681fcb7c24a73ac26fa8b97a6e909/src/model/immutable/EditorBidiService.js#L27) at every [EditorState.push](https://github.com/facebook/draft-js/blob/1ce48f67728681fcb7c24a73ac26fa8b97a6e909/src/model/immutable/EditorState.js#L348); I think the perfs can be improved by:
    - Memoize the calls to `direction`
    - Only update the entries in `directionMap` for blocks that changed (by comparing a the new document with the old one)
- I've had to implement a `State.setDocument` so that all updates of document trigger an update of `directionMap`, Let me know if you still prefer to use `.merge({ document })` directly and I'll find another solution

ianstormtaylor commented:
@SamyPesse this is awesome!!! Going through now and reviewing!

ianstormtaylor commented:
I think we can make a few changes that will simplify things:

1. I think we should change `getDirectionMap()` to be a method on the `Document` model (and memoized). Ideally it would be recursive too, checking `getDirection(default)` on the `Block` model (which is passed the `lastValue`). That way we'd eliminate the need to keep it updated on `State`, but still have performance handled due to memoization.

2. As currently written is seems like this `textAlignment` property is just convenience for setting the `text-align` CSS property? What's the use case when that would be used? Off the top of my head, it seems like we're going to be determining the direction on a per-block basis, but if they wanted to prescribe that text alignment should always be right-aligned, they could add that themselves? Basically, no need for us to special case it, when they can just use the `style=` property instead.

3. Should the direction changing be handled in `shouldComponentUpdate` for nodes?

ianstormtaylor commented on lib/utils/bidi.js:
```diff
@@ -0,0 +1,51 @@
+
+    let bidiMap = new OrderedMap(map)
+
+    if (prevBidiMap != null && is(prevBidiMap, bidiMap)) {
```
We might not need this after memoizing, not sure.

SamyPesse commented:
@ianstormtaylor I've made two changes:

- I've removed the `textAlignment` prop
- I've removed the `directionMap` on State and replaced it by a `node.textDir`, the call to `direction(text)` is memoized (Do I need to implement a limitation on the memoized ? I've seen that there is none on `utils/memoize.js`)

ianstormtaylor commented:
Nice! Could you remove the old `setDocument`-related logic, and then I can help merge it in. I think I'll change `.textDir` to be `.direction`, and I'll figure out what we need to do for memoization logic!

SamyPesse commented:
@ianstormtaylor Done!

ianstormtaylor commented:
Awesome, thanks!

ianstormtaylor commented:
Published in `0.10.0` as well, thanks!
